### ISSUE TITLE
"NoSuchMethodError: method not found: 'call'" in pre-built app

### ISSUE THREAD

**Issue by DanTup** [flutter/devtools#193]
This happens when I launch the tools (so possibly the inspector view?). It only happens in the pre-built version, not if I run `webdev serve` (so maybe caused by a difference between DDC/dart2js?).

```
[{"event":"app.inited"}]
main.dart.js:1358 Uncaught Error: NoSuchMethodError: method not found: 'call'
Receiver: Closure 'minified:D5'
Arguments: ["width", 35, 6, null]
    at Object.k (main.dart.js:1358)
    at D5.dart.h.it (main.dart.js:5849)
    at Object.J.p9 (main.dart.js:18364)
    at Object.lG (main.dart.js:1321)
    at Object.wz (main.dart.js:1332)
    at Object.wB (main.dart.js:1327)
    at Object.l6 (main.dart.js:5173)
    at dart.BP (main.dart.js:7611)
    at main.dart.js:7606
    at w (split.min.js:2)
```

![screenshot 2019-01-30 at 9 42 28 am](https://user-images.githubusercontent.com/1078012/51972526-79835580-2473-11e9-9b6a-83920eae3c9d.png)

**Comment by devoncarew** [comment]
I wonder if we should build the dart2js version of the app with less minification / obfuscation?

And or perhaps have a way to unwind minified stack traces? We'd need to publish some of the dart2js metadata files likely to do this, and make sure that exception reports had the version that was published.

**Comment by DanTup** [comment]
Using a `build.yaml` to turn set `-O0` the error is slightly better:

```
main.dart.js:5096 Uncaught Error: NoSuchMethodError: method not found: 'call'
Receiver: Closure 'flexSplit_closure0'
Arguments: ["width", 35, 6, null]
    at Object.dart.wrapException (main.dart.js:5096)
    at flexSplit_closure0.dart.Object.noSuchMethod$1 (main.dart.js:20666)
    at Object.J.noSuchMethod$1$ (main.dart.js:60382)
    at Object.dart.Primitives_functionNoSuchMethod (main.dart.js:4895)
    at Object.dart.Primitives__genericApplyFunctionWithPositionalArguments (main.dart.js:4941)
    at Object.dart.Primitives_applyFunctionWithPositionalArguments (main.dart.js:4930)
    at Object.dart.Primitives_applyFunction (main.dart.js:4901)
    at Object.dart.Function__apply1 (main.dart.js:18809)
    at Object.dart.Function_apply (main.dart.js:18803)
    at dart._callDartFunctionFast (main.dart.js:31713)
``

The stack if I break on caught exceptions looks like this:

![screenshot 2019-01-30 at 5 05 38 pm](https://user-images.githubusercontent.com/1078012/51998838-55df0000-24b1-11e9-81b3-311a05367eb7.png)

My guess is it's related to this `allowInterop` call?

```dart
return _split(
    parts.map((o) => o is CoreElement ? o.element : o).toList(),
    _SplitOptions(
      elementStyle: allowInterop((dimension, size, gutterSize) {
        return js_util.jsify({
          'flex-basis': 'calc($size% - ${gutterSize}px)',
        });
      }),
```

**Comment by devoncarew** [comment]
> Using a build.yaml to turn set -O0 the error is slightly better

Do you have a sense for the size difference in the compiled JS output? This would be a good flag to use, at least for a while until we're able to un-minify stack traces well.

@jacob314 can speak to the JS interop question - he implemented much of it :)

**Comment by DanTup** [comment]
1.7MB for -O1 and 2.6MB for -O0 (that's just main.dart.js, I presume if there are other scripts it may be similar).

I don't know how that translates in the real world though, I'd guess gzip will compress the larger one much more, though I don't know how it'll affect parsing.

I think using the unoptimised version for now (while in dev/preview) is reasonable to help track issues down. We need to create a `build.yaml` file in order to set it though, doesn't seem to be a switch we can pass to `webdev`.

**Comment by jacob314** [comment]
The unoptimized version sounds fine for now. I expect we will only need to switch if there is a noticeable startup performance difference.

That error makes me think that the JS library isn't being loaded. If you look on the network pane do you see that the splitter.js is loading successfully? Do you have precise instructions to repro?

**Comment by DanTup** [comment]
> That error makes me think that the JS library isn't being loaded.

I assumed all the code was being compiled into the same main.dart.js (because it's 2.6MB) but I don't know it that is a valid assumption!

> Do you have precise instructions to repro?

If you build the app (`pub global run webdev build`) then serve it using the local script (`dart bin/devtools.dart`) and open in your browser, you should see it (you may need to add a port to the querystring to have it connect to a VM, I can't remember).

I can dig more into it tomorrow if you don't get to it.

**Comment by DanTup** [comment]
> Dart2js doesn't have a way to bundle in js libraries

Aha, sorry - I misunderstood the code here. I thought the split was our local code, but I see we're pulling in an external library (split.js). I see what's happening now - we don't have the `third_party` folder in the build output. I can easily copy it for now to fix this, but the build script I made is getting messier and messier, so we'll probably want a better way to build the package!

**Comment by DanTup** [comment]
Scratch that, `third_party` is in the `web` folder so it is in the build output, and it is being loaded in the browser (I can see in the devtools a 200 response and the JS lib). So back to square one :-)

**Comment by DanTup** [comment]
Ok, I think I tracked this down with the help of https://github.com/dart-lang/sdk/issues/31964. It happens when you call a function with the wrong number of args. I replaced splitjs with an unminified version and this is in the call stack:

![screenshot 2019-01-31 at 10 49 23 am](https://user-images.githubusercontent.com/1078012/52049496-e49d5c80-2545-11e9-9dab-4d8490dec4b4.png)

Our `_ElementStyleCallback` only has three args (which is what the docs for splitjs shows), but the code is passing a fourth. I'm not sure why this doesn't fail under DDC though (the issue linked above suggests they should be consistent).

Adding an extra arg to our callbacks (`int index`) seems to solve the issue (and still works under DDC). I guess we need to be careful with external code (eg. when updating) that they don't make changes that are non-breaking for JS, but breaking for us.

I'll open a PR with the fix that slow re-enables the dart2js tests on Travis to confirm it's all good there.
